### PR TITLE
Part2 find samples and test cleanup

### PR DESCRIPTION
**Pull request by iamigo**
use findSamples and clean up tests

package.json
- improve readability with new "testcov" and "postcov" scripts (called from "test" script)
- add simple utility packages for safe set/get

src/heartbeat/heartbeat.js
- reuse httpUtils.doPost instead of setting up the request manually here

src/heartbeat/listener.js
- improve readability

src/heartbeat/utils.js
- when adding or updating a generator, add a "refocus" attribute to the generator, and set the refocus.url and refocus.proxy -- that way we don't have to deal with looking up in global config object from everywhere deep in the stack, which makes all testing much cleaner
- in setupRepeater fn, pass the functions through to createGeneratorRepeater (this makes testing easier)
- minor cleanup for readability, logging, debug
- added a FIXME comment to come back to after we deal with bulk=false
- adopt some of the new queue fns (updateSize, updateFlushTimeout, etc.)

src/remoteCollection/collect.js
- rename the old "collect" fn to "prepareRemoteRequest", then wrap that inside a new "collect" fn which *first* looks up the list of subjects from Refocus, *then* calls prepareRemoteRequest, which calls sendRemoteRequest

src/remoteCollection/handleCollectResponse.js
- added a FIXME comment to come back to after we deal with bulk=false
- pulled the individual sample validation out of the queue enqueue fn so doing it here first before enqueuing
- don't Promise.reject(...) anywhere if we're not handling the reject!!!

src/repeater/repeater.js
- debug and logger cleanup
- rename stopAllRepeat to stopAllRepeaters
- createGeneratorRepeater take functions as args (collect and onProgress) which makes testing easier

src/utils/queue.js
- renamed from queueUtils.js
- expose these functions to improve testability: create, exists, updateSize, get, updateFllushTimeout, enqueue, flushAll

test/sgt.json
- checking in a sample generator template which a bunch of tests use

tests/*
- lots of nock fixes
- add nock for all remote calls to refocus
- silence logging

### CODE REVIEW

File: src/utils/httpUtils.js [hunk 1]
```diff
@@ -85,9 +86,50 @@ function doBulkUpsert(url, userToken, proxy, arr) {
...
+function findSubjects(url, token, proxy, qry) {
+  debug('findSubjects(url=%s, token=%s, proxy=%s, qry=%s)', url,
+    token ? 'HAS_TOKEN' : 'MISSING', proxy, qry);
+  if (!url) {
+    const e = new ValidationError('Missing refocus url');
+    logger.error('findSubjects', e.message);
+    return Promise.reject(e);
+  }
+
+  if (!qry) {
```
Comment (shriramshankar) [inline src/utils/httpUtils.js]: subjectQuery is not a required field on the generator model. If we make it required, do we still have to pass in a queryParam to get all the subjects ?

File: test/heartbeat/heartbeat.js [hunk 2]
```diff
@@ -43,42 +45,45 @@ const generator1Updated = {
...
+  token: 'mygeneratorusertoken',
   interval: 12000,
 };
 
 const generator2 = {
   name: 'generator2_heartbeat',
   interval: 6000,
   aspects: [{ name: 'A2', timeout: '1m' }],
+  subjectQuery: '?absolutePath=S1.S2',
   subjects: [{ absolutePath: 'S1.S2', name: 'S2' }],
```
Comment (shriramshankar) [inline test/heartbeat/heartbeat.js]: We dont need the "subjects" attribute here anymore, correct?

File: src/utils/httpUtils.js (hunk 1 above)
Comment (iamigo) [reply src/utils/httpUtils.js]: When we were taking a list of subjects or a subject query, one or the other was required; once we update the model in Refocus db and api to get rid of the array of subjects, the subject query field should be required. If there is ever a legit use case for generating samples for *all* subjects, then they can use something like "?absolutePath=*", but I'd rather enforce that people think about selecting the "right" subjects

File: test/heartbeat/heartbeat.js (hunk 2 above)
Comment (iamigo) [reply test/heartbeat/heartbeat.js]: We don't, but can we tackle cleaning that up when we handle the "bulk=false" story?

**Pull request merged**